flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/RangePartitioner.java

```diff
@@ -0,0 +1,110 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.sink.shuffle;
+
+import java.util.Random;
+import java.util.concurrent.atomic.AtomicLong;
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.common.functions.Partitioner;
+import org.apache.flink.table.data.RowData;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.SortOrder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** The wrapper class */
+@Internal
+public class RangePartitioner implements Partitioner<StatisticsOrRecord> {
+  private static final Logger LOG = LoggerFactory.getLogger(RangePartitioner.class);
+
+  private final Schema schema;
+  private final SortOrder sortOrder;
+
+  private transient AtomicLong roundRobinCounter;
+  private transient Partitioner<RowData> delegatePartitioner;
+
+  public RangePartitioner(Schema schema, SortOrder sortOrder) {
+    this.schema = schema;
+    this.sortOrder = sortOrder;
+  }
+
+  @Override
+  public int partition(StatisticsOrRecord wrapper, int numPartitions) {
+    if (wrapper.hasStatistics()) {
+      this.delegatePartitioner = delegatePartitioner(wrapper.statistics());
+      return (int) (roundRobinCounter(numPartitions).getAndIncrement() % numPartitions);
+    } else {
+      if (delegatePartitioner != null) {
+        return delegatePartitioner.partition(wrapper.record(), numPartitions);
+      } else {
+        int partition = (int) (roundRobinCounter(numPartitions).getAndIncrement() % numPartitions);
+        LOG.trace("Statistics not available. Round robin to partition {}", partition);
+        return partition;
+      }
+    }
+  }
+
+  private AtomicLong roundRobinCounter(int numPartitions) {
+    if (roundRobinCounter == null) {
+      // randomize the starting point to avoid synchronization across subtasks
+      this.roundRobinCounter = new AtomicLong(new Random().nextInt(numPartitions));
+    }
+
+    return roundRobinCounter;
+  }
+
+  private Partitioner<RowData> delegatePartitioner(GlobalStatistics statistics) {
+    if (statistics.type() == StatisticsType.Map) {
+      return new MapRangePartitioner(schema, sortOrder, statistics.mapAssignment());
+    } else if (statistics.type() == StatisticsType.Sketch) {
+      return new SketchRangePartitioner(schema, sortOrder, statistics.rangeBounds());
+    } else {
+      throw new IllegalArgumentException(
+          String.format("Invalid statistics type: %s. Should be Map or Sketch", statistics.type()));
+    }
+  }
```

Comment:
I still struggling a bit creating a good mental model for the `GlobalStatistics` distribution.

I feel that separating out the `GlobalStatistics` would be good if we could create a data structure which could be used by both `MapRangePartitioner` and `SketchRangePartitioner`, but sadly these partitioners require different input data.

Would it make sense to serialize/parametrize the partitioner on the JM side and send it in the records instead of the `GlobalStatistics` in the `StatisticsOrRecord`?
Like `PartitionerOrRecord`, and on the partitioner side we just call the `partition` and forget it.

Would this be easier to understand for others?

Reply:
> I feel that separating out the GlobalStatistics would be good if we could create a data structure which could be used by both MapRangePartitioner and SketchRangePartitioner

`GlobalStatistics` is for that purpose. it contains either map assignment or range bounds.

> Would it make sense to serialize/parametrize the partitioner on the JM side and send it in the records instead of the GlobalStatistics in the StatisticsOrRecord?

not sure if there is much benefit to serialize the delegate partitioner on the JM side and ship to the subtasks. that would require Java serialization and make checkpoint state more complex for schema evolution.

Reply:
My issue with the GlobalStatistics is that it is not really global. The main payload (mapAssignments and rangeBounds) are different for the different Partitioners. So basically we pretend that we send a statistics, but in reality the data already describes the partitioner. Maybe it would be cleaner to accept this, and mirror this change in the code/class names etc.

Instead of writing a serializer for the GlobalStatistics, we can write a serializer for the respective partitioners. I don't see this as a blocker.

I don't have a very strong opinion on this, and I wanted you to understand why I feel that the code is a bit awkward in this case. What is usually a yellow flag for me, is to have an object where some fields are mutually exclusive. I try to examine those again to understand if we have a single object, or we just merged 2 objects to a single one.

Reply:
`GlobalStatistics` is global in the sense it is aggregated statistics by coordinator. We have a facade/proxy partitioner that would delegate the partition decision to underline partitioner based on the statistics type.

Reply:
After our offline discussion I understand your points better. I understand that the concept is
- That we collect the statistics
- Then use the statistics to create a partitioner
and we try to stick to the concept.

I still think it is confusing, that currently we collect 2 types of statistics and each of them are tightly coupled to the 2 types of partitioners we have. (`Type.Map` is always used by `MapRangePartitioner`, and `Type.Sketch` is always used by `SketchRangePartitioner`). We could create the `MapRangePartitioner` and the `SketchRangePartitioner` on the coordinator side, and send them to a PartitionerExecutor which just deserializes them and runs them. So the real logic would be in a single place (`DataStatisticsCoordinator`).

That said, the improvement is just coding style preference, as the Partitioners still need to serialize the underlying statistics, so the performance would be the same. 

So we can move forward with your proposed solution.
Thanks for the discussion!